Refactor GP module

fehiepsi opened the pull request:
This pull request refactors GP module in favor of #765. Lots of codes are removed in models, the tutorial is updated to newest changes. I also make docs look better. Especially, use Parameterized class introduced at #765 , we can
- Fix a param (so it will not be learned)
- Set prior to a param
- Set constraint to a param.

The only problem I got is I can't replicate the result of the Sparse model in the tutorial. I have checked things carefully, the problem is not lying at this refactoring, but at Pyro's internal changes. I checked each pull request and saw that the behavior of tutorial is different after #695 get merged. I guess there is some things change in that pull request?

@fritzo

eb8680 commented:
@fehiepsi I can help troubleshoot, can you explain what you think is going wrong?

fehiepsi commented:
@eb8680 Thanks! In the current dev version of [GP tutorial](https://github.com/uber/pyro/blob/dev/tutorial/source/gp.ipynb), at sparse approximation part, we can see that variance is small at inducing points and large at the other points. When I run the tutorial again with the dev branch, the plot is similar to what it looks in this pull request (we can't recognize the differences of variances between inducing points and remaining points). So I checkouted each commit from git log, installed Pyro again and again to see which commit made things change. It turns out that the change happened at the pull request #695. I am not sure that things work better or not after that pull request. I just can't replicate the old behavior. Maybe the inference works better now (because the model learnt the mean pretty well!!).

eb8680 commented:
Thanks, I'll look into it, things shouldn't have changed.

fehiepsi commented:
@eb8680 I am really sorry if I have wasted your time to diagnose the problem. After checking the refactor of Poutine, I think that there is nothing change on the inference due to that pull request. It's a great refactoring (thanks to it, I understand more about the internal process of Pyro). The problem lies at the random nature of stochastic inference. I just gave a random seed at the beginning of the notebook, so when I delete some cells, change num_steps,... the later inference will change. I can replicate the old behavior of SGPR with some "random" changes (num_steps of GPR, which are run before SGPR in the notebook).

Reason for the good prediction of the mean: Before this pull request, I missed a `trace_term` for SparseMultivariateNormal', so VFE is just the same as DTC. With this pull request, we can see that VFE performs pretty well.

The problem is solved! Is there someone would like to review this?

fritzo commented on pyro/contrib/gp/__init__.py:
```diff
@@ -1,15 +1,136 @@
+        default value.
+
+        :param str param: Name of a parameter.
+        :param torch.Tensor value: A tensor to be fixed to ``param``.
```
PyTorch tensors will soon be deprecated. Could you change the input to be a detached Variable?
```py
def fix_param(self, param, value=None):
    if value is None:
        value = getattr(self, param).detach()
    self._fixed_params[param] = value
```

fehiepsi replied:
yes, a detached variable doesn't require grad, which is the behavior I want. Thanks

fritzo commented on pyro/contrib/gp/__init__.py:
```diff
@@ -1,15 +1,136 @@
+            param_name = pyro.param_with_module_name(self._name, param)
+
+        if prior is None:
+            constraint = self._constraints[param] if param in self._constraints else None
```
nit: In a few places you can simplify conditional dict lookup using `.get()`:
```py
constraint = self._constraints.get(param)
```

fritzo commented on pyro/contrib/gp/__init__.py:
```diff
@@ -1,15 +1,136 @@
+            raise ValueError("Mode should be either 'model' or 'guide', but got {}.".format(mode))
+        for param in self._parameters:
+            self._register_param(param, mode)
+        return self
```
nit: It's kind of weird to `return self` here. I realize some codebases use method chaining heavily, but the patter is seldom used in Pyro.

fehiepsi replied:
I agree, it just provides me a shortcut for `kernel = self.kernel.set_mode(...)`, which is unnecessary.

fritzo commented on pyro/contrib/gp/__init__.py:
```diff
@@ -1,15 +1,136 @@
+            value = getattr(self, param).data
+        self._fixed_params[param] = Variable(value)
+
+    def unfix_param(self, param):
```
Is there an actual use case for this? If there's nothing immediate, maybe it's simpler to add it when we need it later on.

fehiepsi replied:
I didn't use it ^^! will remove it soon

fritzo commented on pyro/contrib/gp/__init__.py:
```diff
@@ -1,15 +1,136 @@
+
+    def set_prior(self, param, prior):
+        """
+        Sets a prior to a parameter. Use `set_prior(param, None)` to remove the prior distribution of
```
Is there a use case for unsetting a prior? If there's no foreseeable use case, it would be better to not support this functionality.

fehiepsi replied:
The use case is when we want to reuse some modules, but are lazy to redefine it. ^^! Because it is not important (in my case, I don't use it), I will delete that part of docs.

fritzo commented on pyro/contrib/gp/__init__.py:
```diff
@@ -1,15 +1,136 @@
+        if param in self._fixed_params:
+            self._registered_params[param] = self._fixed_params[param]
+            return
+        prior = self._priors[param] if param in self._priors else None
```
```py
prior = self._priors.get(param)
```

fritzo commented on pyro/contrib/gp/__init__.py:
```diff
@@ -1,15 +1,136 @@
+    <http://pytorch.org/docs/master/nn.html#torch.nn.Module>`_ for more information.
+    """
+
+    def __init__(self):
```
nit: What do you think of adding `name` as an `__init__` arg rather than setting `self._name = ...` in derived classes? This seems a little safer to me.

fehiepsi commented:
@fritzo I have addressed your reviews. There is an additional change: when we call `set_mode()` method, it also automatically sets `mode` for submodules which are derived from `Parameterize` class. This behavior is similar to other Pytorch module's behaviors (`.train()`, `.eval()`, `.double()`, `.cuda()`,... So in the model, we just simply call `self.set_mode(...)` instead of several calls for its submodule (kernel, likelihood).

fritzo commented on pyro/contrib/gp/__init__.py:
```diff
@@ -1,15 +1,132 @@
+            p = pyro.sample(param_name, prior)
+        else:  # prior != None and mode = "guide"
+            MAP_param_name = param_name + "_MAP"
+            MAP_param_0 = Variable(prior.torch_dist.mean.data.clone(), requires_grad=True)
```
This should be
```diff
- prior.torch_dist.mean
+ prior.analytic_mean()
```
but I'll address that in a follow-up PR.
